File: Application/src/main/java/com/shareyourproxy/api/domain/model/SharedLink.java
package com.shareyourproxy.api.domain.model;

import android.os.Parcelable;

import com.shareyourproxy.api.gson.AutoGson;

import java.util.UUID;

import auto.parcel.AutoParcel;

/**
 * Upload information to firebase to create shared links for group channels.
 */
@AutoParcel
@AutoGson(autoValueClass = AutoParcel_SharedLink.class)
public abstract class SharedLink implements Parcelable {

    /**
     * Create a new Contact.
     *
     * @param user of shared content
     * @param group to share
     * @return shared link
     */
    @SuppressWarnings("unused")
    public static SharedLink create(User user, Group group) {
        String id = UUID.randomUUID().toString();
        return builder().id(id).userId(user.id().value()).groupId(group.id().value()).build();
    }

    /**
     * Contact builder.
     *
     * @return this Contact.
     */
    public static Builder builder() {
        return new AutoParcel_SharedLink.Builder();
    }

    /**
     * Group Id.
     *
     * @return name
     */
    public abstract String id();

    /**
     * Group to share.
     *
     * @return name
     */
    public abstract String groupId();

    /**
     * Shared Group's User.
     *
     * @return name
     */
    public abstract String userId();

    @AutoParcel.Builder
    public interface Builder {

        /**
         * Set the Id.
         *
         * @param id group id
         * @return builder
         */
        Builder id(String id);

        /**
         * Set the group.
         *
         * @param group group object
         * @return builder
         */
        Builder groupId(String group);

        /**
         * Set the shared links user.
         *
         * @param user of shared group
         * @return builder
         */
        Builder userId(String user);

        /**
         * BUILD.
         *
         * @return SharedLink
         */
        SharedLink build();
    }
}

File: Application/src/main/kotlin/com/shareyourproxy/api/domain/model/SharedLink.kt
package com.shareyourproxy.api.domain.model

import android.os.Parcel
import android.os.Parcelable
import com.shareyourproxy.util.BaseParcelable
import java.util.*

/**
 * Upload information to firebase to create shared links for group channels.
 */
internal data class SharedLink(val id: String, val userId: String, val groupId: String) : BaseParcelable {
    public constructor(userId: String, groupId: String) : this(UUID.randomUUID().toString(), groupId, userId)

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(id)
        dest.writeString(userId)
        dest.writeString(groupId)
    }

    companion object {
        val CREATOR = object : Parcelable.Creator<SharedLink> {
            override fun createFromParcel(parcel: Parcel) = readParcel(parcel)
            override fun newArray(size: Int): Array<SharedLink?> = arrayOfNulls(size)
        }

        private fun readParcel(parcel: Parcel) = SharedLink(parcel.readString(), parcel.readString(), parcel.readString())
    }
}
